File: 2025-9322-team04_finproject_python/Client_Python/player/controller/GameProperController.py
import threading
import time
from Client_Python.manager.PlayerCorbaManager import PlayerCorbaManager
from Client_Python.player.main import GameProperMain
from Client_Python.player.main.HomepageMain import HomepageMain
from Client_Python.player.model.GameProperModel import GameProperModel

class GameProperController:
    def __init__(self, username, current_round, game_id, actual_word):
        orb_args = ["-ORBInitRef", "NameService=corbaloc::localhost:4321/NameService"]
        self.manager = PlayerCorbaManager.get_instance(orb_args)
        self.model = GameProperModel(actual_word)
        self.game_id = game_id
        self.username = username
        self.has_submitted = False
        self.signal_next_round = False
        self.next_round_started = False
        self.current_round = current_round
        self.max_rounds = 3
        self.setup_timer()

    def setup_timer(self):
        def game_timer_loop():
            while self.model.get_timer_seconds() > 0 and not self.has_submitted and self.model.get_lives() > 0:
                time.sleep(1)
                self.model.decrement_timer()
               # print(
                   # f"Time Remaining: {self.model.get_timer_seconds()}")

            if not self.has_submitted and (self.model.get_timer_seconds() <= 0 or self.model.get_lives() <= 0):
                self.has_submitted = True
                self.notify_player_done()
                print("Waiting for others...")

        timer_thread = threading.Thread(target=game_timer_loop, daemon=True)
        timer_thread.start()

    def handle_guess(self, guess):
        if self.has_submitted or self.model.get_lives() <= 0:
            return
        if len(guess) != 1 or not guess.isalpha():
            print("Please enter a single alphabetic character.")
            return

        print(f"You entered {guess}")
        is_correct = self.model.is_letter_in_word(guess)
        if is_correct:
            print("Your guess is correct")
        else:
            print("Your guess is wrong")
            self.model.decrement_lives()

        self.update_revealed_letters(guess)

        if not self.has_submitted and self.model.is_word_fully_revealed():
            self.has_submitted = True
            self.notify_player_done()
            print("You have completed the word.\nWaiting for others...")

        if not self.has_submitted and (self.model.get_timer_seconds() <= 0 or self.model.get_lives() <= 0):
            self.has_submitted = True
            self.notify_player_done()
            print("Time or lives exhausted.\nWaiting for others...")

    def update_revealed_letters(self, guess):
        actual_word = self.model.get_word_to_guess()
        revealed = self.model.get_revealed_letters()
        for i, ch in enumerate(actual_word):
            if ch == guess:
                revealed[i] = True

    def notify_player_done(self):
        try:
            guessed_correctly = self.model.is_word_fully_revealed()
            game_service = self.manager.get_game_service()
            server_round_number = game_service.getCurrentRound(self.game_id)
            game_service.setPlayerDone(self.username, server_round_number, guessed_correctly)
            game_service.advanceToNextRoundIfReady(self.username, server_round_number)
            self.check_all_players_done()
        except Exception as e:
            print(f"Error notifying server: {e}")

    def check_all_players_done(self):
        def wait_loop():
            while True:
                try:
                    game_service = self.manager.get_game_service()
                    if game_service.isGameOver():
                        self.show_final_scores_and_return_home()
                        break

                    server_round = game_service.getCurrentRound(self.game_id)
                    if server_round > self.current_round:
                        self.proceed_to_next_round()
                        break

                    if game_service.areAllPlayersDone(self.username, self.current_round):
                        if self.current_round >= self.max_rounds - 1:
                            self.end_game()
                        else:
                            self.show_final_word_and_countdown()
                        break
                except Exception as ex:
                    print(f"Error checking players: {ex}")
                time.sleep(1)

        threading.Thread(target=wait_loop, daemon=True).start()

    def show_final_word_and_countdown(self):
        countdown_seconds = 5

        def countdown_timer():
            seconds_left = countdown_seconds
            while seconds_left > 0:
                actual_word = self.model.get_word_to_guess()
                revealed = self.model.get_revealed_letters()
                masked_word = ''.join([ch if revealed[i] else '_' for i, ch in enumerate(actual_word)])
                print(f"The word was: {masked_word}\nNext round in {seconds_left} seconds")
                time.sleep(1)
                seconds_left -= 1
            self.wait_for_next_round_on_server()

        threading.Thread(target=countdown_timer, daemon=True).start()

    def wait_for_next_round_on_server(self):
        def check_server():
            while True:
                try:
                    game_service = self.manager.get_game_service()
                    server_round = game_service.getCurrentRound(self.game_id)
                    if server_round > self.current_round:
                        self.proceed_to_next_round()
                        break
                except Exception as ex:
                    print(f"Error waiting for next round: {ex}")
                time.sleep(1)

        threading.Thread(target=check_server, daemon=True).start()

    def proceed_to_next_round(self):
        if self.next_round_started:
            return
        self.next_round_started = True
        self.has_submitted = False
        self.signal_next_round = True
        #next_round = self.current_round + 1

        #GameProperMain(next_round, self.username)

    def end_game(self):
        print("Game Over!\nThe final word was: " + self.model.get_word_to_guess())
        print("Thank you for playing!")

    def show_final_scores_and_return_home(self):
        try:
            game_service = self.manager.get_game_service()
            scores = game_service.getScoreSummary()
            winner = game_service.getWinner()
            print("Game Over!\nWinner: " + winner + "\n\n" + scores)
            HomepageMain(self.username)
        except Exception as ex:
            print(f"Error retrieving final scores: {ex}")

#--------functions to pass to Main------------

#For display_word
    def pass_revealed_letters(self):
        return self.model.get_revealed_letters()
#For lives and timer
    def pass_timer(self):
        return self.model.get_timer_seconds()

    def pass_lives(self):
        return self.model.get_lives()

File: 2025-9322-team04_finproject_java/src/main/java/Client_Java/player/controller/GameProperController.java
package Client_Java.player.controller;

import Client_Java.player.model.GameProperModel;
import Client_Java.player.model.HomepageModel;
import Client_Java.player.model.PlayerLoginModel;
import Client_Java.player.view.GameProperView;
import Client_Java.player.view.HomepageView;
import Client_Java.player.PlayerCorbaManager;
import Client_Java.player.view.PlayerLoginView;
import compilations.*;
import javax.swing.*;

public class GameProperController {
    private final String gameId;
    private final String username;
    private final GameProperModel model;
    private final GameProperView view;
    private final PlayerCorbaManager corbaManager;

    private Timer gameTimer;
    private boolean hasSubmitted = false;
    private boolean nextRoundStarted = false;

    private int currentRound;
    private final int maxRounds = 3;

    public GameProperController(GameProperModel model, GameProperView view,
                                PlayerCorbaManager corbaManager,
                                String username, int currentRound, String gameId) {
        this.model = model;
        this.view = view;
        this.corbaManager = corbaManager;
        this.username = username;
        this.currentRound = currentRound;
        this.gameId = gameId;

        view.setRound(currentRound);
        setupKeyboard();
        setupTimer();
        setupHeaderButtons();
        startSessionValidationTimer();
    }

    private void setupKeyboard() {
        for (char c = 'A'; c <= 'Z'; c++) {
            JButton key = view.getKeyButton(c);
            if (key != null) {
                final char guessChar = c;
                key.addActionListener(e -> handleGuess(guessChar, key));
            }
        }
    }

    private void handleGuess(char guess, JButton key) {
        if (hasSubmitted || model.getLives() <= 0) return;

        key.setEnabled(false);
        boolean isCorrect = model.isLetterInWord(guess);

        if (isCorrect) {
            updateRevealedLetters();
        } else {
            model.decrementLives();
            updateHearts();
        }

        updateRevealedLetters();

        if (!hasSubmitted && model.isWordFullyRevealed()) {
            hasSubmitted = true;
            gameTimer.stop();
            notifyPlayerDone();
            view.showWaitingForOthersDialog();
        }

        if (!hasSubmitted && (model.getTimerSeconds() <= 0 || model.getLives() <= 0)) {
            hasSubmitted = true;
            gameTimer.stop();
            notifyPlayerDone();
            view.showWaitingForOthersDialog();
        }
    }

    private void updateRevealedLetters() {
        boolean[] revealed = model.getRevealedLetters();
        JLabel[] labels = view.getLetterLabels();
        for (int i = 0; i < labels.length; i++) {
            if (revealed[i]) {
                labels[i].setText(String.valueOf(model.getWordToGuess().charAt(i)));
            }
        }
    }

    private void updateHearts() {
        JLabel[] hearts = view.getHeartLabels();
        int livesLeft = model.getLives();
        for (int i = 0; i < hearts.length; i++) {
            hearts[i].setVisible(i < livesLeft);
        }
    }

    private void setupTimer() {
        gameTimer = new Timer(1000, e -> {
            model.decrementTimer();
            view.getTimerLabel().setText("Time Remaining: " + model.getTimerSeconds());

            if (!hasSubmitted && (model.getTimerSeconds() <= 0 || model.getLives() <= 0)) {
                gameTimer.stop();
                hasSubmitted = true;
                notifyPlayerDone();
                view.showWaitingForOthersDialog();
            }
        });
        gameTimer.start();
    }

    private void setupHeaderButtons() {
        view.getMinimizeBtn().addActionListener(e -> view.getFrame().setState(JFrame.ICONIFIED));
        view.getMaximizeBtn().addActionListener(e -> {
            JFrame frame = view.getFrame();
            if (frame.getExtendedState() == JFrame.MAXIMIZED_BOTH) {
                frame.setExtendedState(JFrame.NORMAL);
            } else {
                frame.setExtendedState(JFrame.MAXIMIZED_BOTH);
            }
        });
        view.getCloseBtn().addActionListener(e -> {
            gameTimer.stop();
            System.exit(0);
        });
    }

    private void notifyPlayerDone() {
        try {
            boolean guessedCorrectly = model.isWordFullyRevealed();
            GameService gameService = corbaManager.getGameService();
            short serverRoundNumber = gameService.getCurrentRound(gameId);

            gameService.setPlayerDone(username, serverRoundNumber, guessedCorrectly);
            boolean roundAdvanced = gameService.advanceToNextRoundIfReady(username, serverRoundNumber);

            checkAllPlayersDone();
        } catch (Exception e) {
            e.printStackTrace();
            JOptionPane.showMessageDialog(view.getFrame(),
                    "Error notifying server: " + e.getMessage(),
                    "Error", JOptionPane.ERROR_MESSAGE);
        }
    }

    private void checkAllPlayersDone() {
        Timer waitTimer = new Timer(1000, null);
        waitTimer.addActionListener(e -> {
            try {
                GameService gameService = corbaManager.getGameService();

                if (gameService.isGameOver()) {
                    waitTimer.stop();
                    showFinalScoresAndReturnHome();
                    return;
                }

                int serverRound = gameService.getCurrentRound(gameId);
                if (serverRound > currentRound) {
                    waitTimer.stop();
                    proceedToNextRound();
                    return;
                }

                if (gameService.areAllPlayersDone(username, (short) currentRound)) {
                    waitTimer.stop();
                    if (currentRound >= maxRounds - 1) {
                        endGame();
                    } else {
                        showFinalWordAndCountdown();
                    }
                }

            } catch (Exception ex) {
                ex.printStackTrace();
            }
        });
        waitTimer.start();
    }


    private void showFinalWordAndCountdown() {
        view.hideWaitingDialogAndEnableInput();
        int countdownSeconds = 5;
        final int[] secondsLeft = {countdownSeconds};

        Timer countdownTimer = new Timer(1000, null);
        countdownTimer.addActionListener(e -> {
            if (secondsLeft[0] > 0) {
                view.getTimerLabel().setText("The word was: " + model.getWordToGuess() + "\nNext round in " + secondsLeft[0]--);
            } else {
                countdownTimer.stop();
                waitForNextRoundOnServer();
            }
        });
        countdownTimer.setInitialDelay(0);
        countdownTimer.start();
    }

    private void waitForNextRoundOnServer() {
        Timer waitForServer = new Timer(1000, null);
        waitForServer.addActionListener(e -> {
            try {
                GameService gameService = corbaManager.getGameService();
                int serverRound = gameService.getCurrentRound(gameId);
                if (serverRound > currentRound) {
                    waitForServer.stop();
                    proceedToNextRound();
                }
            } catch (Exception ex) {
                ex.printStackTrace();
            }
        });
        waitForServer.start();
    }

    private void proceedToNextRound() {
        if (nextRoundStarted) return;

        nextRoundStarted = true;
        hasSubmitted = false;
        int nextRound = currentRound + 1;

        try {
            WordService wordService = corbaManager.getWordService();
            String newWord = wordService.getCurrentWord(gameId);
            if (newWord == null || newWord.trim().isEmpty()) {
                newWord = "FALLBACK";
            }

            view.getFrame().dispose();

            GameProperModel nextModel = new GameProperModel(newWord);
            GameProperView newView = new GameProperView(newWord);
            new GameProperController(nextModel, newView, corbaManager, username, nextRound, gameId);

        } catch (Exception e) {
            e.printStackTrace();
            JOptionPane.showMessageDialog(view.getFrame(),
                    "Error proceeding to next round: " + e.getMessage());
        }
    }

    private void endGame() {
        view.hideWaitingDialogAndEnableInput();
        JOptionPane.showMessageDialog(view.getFrame(),
                "Game Over!\nThe final word was: " + model.getWordToGuess() +
                        "\nThank you for playing!");
        view.getFrame().dispose();
    }

    private void showFinalScoresAndReturnHome() {
        view.hideWaitingDialogAndEnableInput();
        try {
            GameService gameService = corbaManager.getGameService();
            String scores = gameService.getScoreSummary();
            String winner = gameService.getWinner();

            JOptionPane.showMessageDialog(view.getFrame(),
                    "Game Over!\nWinner: " + winner + "\n\n" + scores,
                    "Final Scores", JOptionPane.INFORMATION_MESSAGE);

            view.getFrame().dispose();

            HomepageModel homepageModel = new HomepageModel();
            homepageModel.setUsername(username);

            HomepageView homepageView = new HomepageView();
            new HomepageController(homepageModel, homepageView);

            homepageView.show();

        } catch (Exception e) {
            e.printStackTrace();
            JOptionPane.showMessageDialog(view.getFrame(),
                    "Error retrieving final scores: " + e.getMessage(),
                    "Error", JOptionPane.ERROR_MESSAGE);
        }
    }

    private void startSessionValidationTimer() {
        Timer sessionCheckTimer = new Timer(5000, e -> {
            try {
                if (!model.isSessionValid()) {
                    String sessionId = corbaManager.getInstance().getSessionId();
                    corbaManager.getInstance().getLoginService().logoutPlayer(sessionId); // Invalidate session

                    JOptionPane.showMessageDialog(view.getFrame(),
                            "Your session has expired or you have been logged out elsewhere.",
                            "Session Expired",
                            JOptionPane.WARNING_MESSAGE);

                    view.getFrame().dispose(); // Close current view
                    PlayerLoginView view = new PlayerLoginView();
                    PlayerLoginModel model = new PlayerLoginModel();
                    new PlayerLoginController(model, view);
                }
            } catch (Exception ex) {
                System.err.println("Error checking session validity: " + ex.getMessage());
            }
        });
        sessionCheckTimer.start();
    }
}

